### ISSUE TITLE
BeforeSuite should run before any Describe blocks

### ISSUE THREAD

**Issue by stevenroose** [onsi/ginkgo#457]
Even though, I read [this comment](https://stackoverflow.com/a/36204239/749521), I don't think that makes sense.

If you have some variables that you need, not only in `It` blocks (whatever the "It" stands for), but also before that, right now the only way is to manually call a function before calling ginkgo. That is literally what the BeforeSuite should be doing.

**Comment by stevenroose** [comment]
For example, I'm not writing a test suite for the package itself, but a REST API test suite. So I have a "test config" file with some parameters. These parameters could even define what exact tests should be ran, so it would be impossible to not call these parameters before `It`.

**Comment by kirederik** [comment]
Hi @stevenroose,

I think you are missing the [BeforeEach](https://onsi.github.io/ginkgo/#extracting-common-setup-beforeeach) block. That is, rather than setting up configuration variables inside the `Describe/Context` block, you wrap it inside a [`BeforeEach`](https://onsi.github.io/ginkgo/#extracting-common-setup-beforeeach) like this:

```
# suite
var serverURL string

var _ = BeforeSuite(func() {
    r = router.NewRouter()
    s = httptest.NewServer(r)
    serverURL = s.URL
})
```

```
# test
var _ = Describe("Handlers/CartContentsHandler", func() {
  var url url.URL

  BeforeEach(func() {
    url = serverURL + "/cart-contents"
  })

  It("Makes a GET request", func() {
    # ....
  })
})

```

**Comment by stevenroose** [comment]
Consider the following suite:
```go
var testConfig = struct {
	AccountsToCheck []string `yaml:"accounts"`
}{}

var _ = BeforeSuite(func() {
	// loadfile
	content := "accounts: \n- one\n- two\n"
	err := yaml.Unmarshal(content, &testConfig)
	Expect(err).To(Succeed())
})

var _ = Describe("test accounts", func() {
	for accIdx, account := testConfig.AccountsToCheck {
		Context(fmt.Sprintf("account #%d (%s)", accIdx, account), func() {
			It("should do something", func() {
				//blabla
			})
		})
	}
})
```

The gist is: **_You can't use any dynamically loaded test parameters to create contexts._** Unless you suggests contexts should be used within `It`s.

**Comment by onsi** [comment]
It does seem the naming is getting in the way.  `BeforeSuite` runs before  any of the tests in the suite run.  The test _tree_, however, is instantiated when the Go program (the test in this case) is loaded into memory and all the top level variables and invocations are run.  The test tree is static and cannot be mutated once the tests begin.  The first thing that happens when the tests begin is the `BeforeSuite` is run.

You can do what you are trying to do by actually moving the contents of the `BeforeSuite` directly into the body of the `Describe` `func`.  If you want to reuse `testConfig` you could have a memorized helper function that you invoke in your `Describe`s to get at your `testConfig`.  Perhaps a bit kludgy but it should get you unblocked.

**Comment by williammartin** [comment]
I'm going to close this out now and if there's any enhancement or recommendation for ginkgo as a result, feel free to reopen or create a new issue.

Thanks for the discussion.
